From 906a473e57f493396f55bf4b41da47cdb6e25ca3 Mon Sep 17 00:00:00 2001
From: zwx1252978 <zhangliang296@h-partners.com>
Date: Sat, 24 Jan 2026 16:48:08 +0800
Subject: [PATCH 1/3] ub: udma: Support for the separate page table feature

driver inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/ID457F
CVE: NA

-----------------------------------------------------------

This patch adds the ability to allow user use separate page table
feature, create a separate page table for hardware access

Signed-off-by: Liang Zhang <zhangliang296@h-partners.com>
---
 src/urma/hw/udma/kernel_headers/udma_abi.h |  4 ++-
 src/urma/hw/udma/udma_u_buf.c              | 33 ++++++++++++++++++----
 src/urma/hw/udma/udma_u_buf.h              | 10 +++----
 src/urma/hw/udma/udma_u_common.h           |  2 ++
 src/urma/hw/udma/udma_u_ctl.c              |  7 ++---
 src/urma/hw/udma/udma_u_db.c               |  5 +++-
 src/urma/hw/udma/udma_u_jfr.c              |  4 +--
 src/urma/hw/udma/udma_u_ops.c              |  1 +
 src/urma/hw/udma/udma_u_segment.c          |  6 ++--
 9 files changed, 51 insertions(+), 21 deletions(-)

diff --git a/src/urma/hw/udma/kernel_headers/udma_abi.h b/src/urma/hw/udma/kernel_headers/udma_abi.h
index 2579627..5d2cad0 100644
--- a/src/urma/hw/udma/kernel_headers/udma_abi.h
+++ b/src/urma/hw/udma/kernel_headers/udma_abi.h
@@ -98,7 +98,8 @@ struct udma_create_ctx_resp {
 	__u32 reduce_enable : 1;
 	__u32 dump_aux_info : 1;
 	__u32 hugepage_enable : 1;
-	__u32 rsv : 20;
+	__u32 sva_sep_mode_en : 1;
+	__u32 rsv : 19;
 	__u32 ue_id;
 	__u32 chip_id;
 	__u32 die_id;
@@ -115,6 +116,7 @@ enum db_mmap_type {
 	UDMA_MMAP_HUGEPAGE,
 	UDMA_MMAP_JFC_PAGE,
 	UDMA_MMAP_JETTY_DSQE,
+	UDMA_MMAP_KERNEL_BUF,
 };
 
 enum {
diff --git a/src/urma/hw/udma/udma_u_buf.c b/src/urma/hw/udma/udma_u_buf.c
index edfe893..e2d1097 100644
--- a/src/urma/hw/udma/udma_u_buf.c
+++ b/src/urma/hw/udma/udma_u_buf.c
@@ -10,6 +10,28 @@
 #include <stdio.h>
 #include "udma_u_buf.h"
 
+void *udma_u_alloc_kernel_buf(struct udma_u_context *ctx, uint32_t buf_size)
+{
+	off_t offset = get_mmap_offset(0, ctx->page_size, UDMA_MMAP_KERNEL_BUF);
+	void *buf;
+	int ret;
+
+	buf = mmap(NULL, buf_size, PROT_READ | PROT_WRITE, MAP_SHARED, ctx->urma_ctx.dev_fd, offset);
+	if (buf == MAP_FAILED) {
+		UDMA_LOG_ERR("mmap failed, buf_size=%u.\n", buf_size);
+		return NULL;
+	}
+
+	ret = madvise(buf, buf_size, MADV_DONTFORK);
+	if (ret) {
+		(void)munmap(buf, buf_size);
+		UDMA_LOG_ERR("buf madvise failed! ret = %d\n", ret);
+		return NULL;
+	}
+
+	return buf;
+}
+
 bool udma_u_alloc_queue_buf(struct udma_u_jetty_queue *q, uint32_t max_entry_cnt,
 			    uint32_t baseblk_size, uint32_t page_size,
 			    bool wrid_en)
@@ -25,7 +47,7 @@ bool udma_u_alloc_queue_buf(struct udma_u_jetty_queue *q, uint32_t max_entry_cnt
 	q->baseblk_mask = q->baseblk_cnt - 1U;
 
 	if (wrid_en) {
-		q->wrid = (uintptr_t *)malloc(q->baseblk_cnt * sizeof(uint64_t));
+		q->wrid = (uintptr_t *)calloc(1, q->baseblk_cnt * sizeof(uint64_t));
 		if (!q->wrid) {
 			UDMA_LOG_ERR("failed to alloc buffer for wrid.\n");
 			return false;
@@ -37,11 +59,11 @@ bool udma_u_alloc_queue_buf(struct udma_u_jetty_queue *q, uint32_t max_entry_cnt
 		if (q->hugepage) {
 			q->qbuf = q->hugepage->va_start;
 		} else {
-			UDMA_LOG_WARN("failed to alloc hugepage buf, switch to alloc normal buf.");
-			q->qbuf = udma_u_alloc_buf(q->qbuf_size);
+			UDMA_LOG_WARN("failed to alloc hugepage buf, switch to alloc normal buf.\n");
+			q->qbuf = udma_u_alloc_kernel_buf(q->ctx, q->qbuf_size);
 		}
 	} else {
-		q->qbuf = udma_u_alloc_buf(q->qbuf_size);
+		q->qbuf = udma_u_alloc_kernel_buf(q->ctx, q->qbuf_size);
 	}
 
 	if (!q->qbuf) {
@@ -130,7 +152,8 @@ static void udma_u_hugepage_del(struct udma_u_context *ctx,
 static struct udma_u_hugepage_priv *
 udma_u_alloc_hugepage_priv(struct udma_u_context *ctx, uint32_t len)
 {
-	off_t offset = get_mmap_offset((UDMA_HUGEPAGE_SIZE / ctx->page_size >> MAP_INDEX_SHIFT), ctx->page_size, UDMA_MMAP_HUGEPAGE);
+	off_t offset = get_mmap_offset((UDMA_HUGEPAGE_SIZE / ctx->page_size >> MAP_INDEX_SHIFT),
+					ctx->page_size, UDMA_MMAP_HUGEPAGE);
 	struct udma_u_hugepage_priv *priv;
 	int ret;
 
diff --git a/src/urma/hw/udma/udma_u_buf.h b/src/urma/hw/udma/udma_u_buf.h
index 59ad210..10ed79e 100644
--- a/src/urma/hw/udma/udma_u_buf.h
+++ b/src/urma/hw/udma/udma_u_buf.h
@@ -13,16 +13,16 @@
 #include <sys/mman.h>
 #include "udma_u_common.h"
 
-bool udma_u_alloc_queue_buf(struct udma_u_jetty_queue *q, uint32_t max_entry_cnt,
-			    uint32_t baseblk_size, uint32_t page_size, bool wrid_en);
-void udma_u_free_queue_buf(struct udma_u_jetty_queue *q);
-void *udma_u_alloc_buf(uint32_t buf_size);
-
 static inline void udma_u_free_buf(void *buf, uint32_t buf_size)
 {
 	(void)munmap(buf, buf_size);
 }
 
+void *udma_u_alloc_kernel_buf(struct udma_u_context *ctx, uint32_t buf_size);
+bool udma_u_alloc_queue_buf(struct udma_u_jetty_queue *q, uint32_t max_entry_cnt,
+			    uint32_t baseblk_size, uint32_t page_size, bool wrid_en);
+void udma_u_free_queue_buf(struct udma_u_jetty_queue *q);
+void *udma_u_alloc_buf(uint32_t buf_size);
 struct udma_u_hugepage *udma_u_alloc_hugepage(struct udma_u_context *ctx, uint32_t len);
 void udma_u_free_hugepage(struct udma_u_context *ctx, struct udma_u_hugepage *hugepage);
 void udma_u_destroy_hugepage(struct udma_u_context *ctx);
diff --git a/src/urma/hw/udma/udma_u_common.h b/src/urma/hw/udma/udma_u_common.h
index c168022..bb9f855 100644
--- a/src/urma/hw/udma/udma_u_common.h
+++ b/src/urma/hw/udma/udma_u_common.h
@@ -40,6 +40,7 @@
 #define UDMA_MIN(x, y) ((x) < (y) ? (x) : (y))
 #define UDMA_MAX(x, y) ((x) > (y) ? (x) : (y))
 #define RTE_SET_USED(x) (void)(x)
+#define UNLIKELY(x) (__builtin_expect(!!(x), 0))
 
 struct udma_u_doorbell {
 	uint32_t id;
@@ -95,6 +96,7 @@ struct udma_u_context {
 	uint32_t		die_id;
 	bool			dump_aux_info;
 	uint32_t		jfr_sge;
+	bool			sva_sep_mode_en;
 	bool			hugepage_enable;
 	pthread_mutex_t		hugepage_lock;
 	struct udma_u_hugepage_priv *hugepage_list;
diff --git a/src/urma/hw/udma/udma_u_ctl.c b/src/urma/hw/udma/udma_u_ctl.c
index 1cbaecd..cd4030a 100644
--- a/src/urma/hw/udma/udma_u_ctl.c
+++ b/src/urma/hw/udma/udma_u_ctl.c
@@ -95,13 +95,12 @@ static int udma_u_init_jfs_queue_buf(struct udma_u_jetty_queue *q,
 		if (q->hugepage) {
 			q->qbuf = q->hugepage->va_start;
 		} else {
-			UDMA_LOG_WARN("failed to alloc hugepage buf, switch to alloc standard buf.");
-			q->qbuf = udma_u_alloc_buf(q->qbuf_size);
+			UDMA_LOG_WARN("failed to alloc hugepage buf, switch to alloc standard buf.\n");
+			q->qbuf = udma_u_alloc_kernel_buf(q->ctx, q->qbuf_size);
 		}
 	} else {
-		q->qbuf = udma_u_alloc_buf(q->qbuf_size);
+		q->qbuf = udma_u_alloc_kernel_buf(q->ctx, q->qbuf_size);
 	}
-
 	if (q->qbuf == NULL) {
 		UDMA_LOG_ERR("failed to alloc queue buffer.\n");
 		free(q->wrid);
diff --git a/src/urma/hw/udma/udma_u_db.c b/src/urma/hw/udma/udma_u_db.c
index 2e8a6b5..0d5c65e 100644
--- a/src/urma/hw/udma/udma_u_db.c
+++ b/src/urma/hw/udma/udma_u_db.c
@@ -100,7 +100,10 @@ static struct udma_u_db_page *udma_add_db_page(struct udma_u_context *ctx,
 		goto err_map;
 
 	db_page->buf.length = page_size;
-	db_page->buf.buf = udma_u_alloc_buf(db_page->buf.length);
+	if (type == UDMA_JFR_PAYLOAD)
+		db_page->buf.buf = udma_u_alloc_buf(db_page->buf.length);
+	else
+		db_page->buf.buf = udma_u_alloc_kernel_buf(ctx, db_page->buf.length);
 	if (!db_page->buf.buf)
 		goto err_buf;
 
diff --git a/src/urma/hw/udma/udma_u_jfr.c b/src/urma/hw/udma/udma_u_jfr.c
index d9a509e..fdd06ca 100644
--- a/src/urma/hw/udma/udma_u_jfr.c
+++ b/src/urma/hw/udma/udma_u_jfr.c
@@ -110,7 +110,7 @@ static int udma_u_alloc_jfr_idx_que(struct udma_u_jfr *jfr)
 	buf_size = align(jfr->wqe_cnt << idx_que->entry_shift,
 			 UDMA_HW_PAGE_SIZE);
 	idx_que->buf.length = align(buf_size, UDMA_HW_PAGE_SIZE);
-	idx_que->buf.buf = udma_u_alloc_buf(idx_que->buf.length);
+	idx_que->buf.buf = udma_u_alloc_kernel_buf(jfr->rq.ctx, idx_que->buf.length);
 	if (!idx_que->buf.buf) {
 		udma_bitmap_free(idx_que->bitmap);
 		idx_que->bitmap = NULL;
@@ -169,6 +169,7 @@ urma_jfr_t *udma_u_create_jfr(urma_context_t *ctx, urma_jfr_cfg_t *cfg)
 		return NULL;
 	}
 
+	udma_jfr->rq.ctx = udma_ctx;
 	udma_u_init_jfr_param(udma_jfr, cfg);
 
 	if (!udma_jfr->lock_free &&
@@ -180,7 +181,6 @@ urma_jfr_t *udma_u_create_jfr(urma_context_t *ctx, urma_jfr_cfg_t *cfg)
 		goto err_alloc_idx;
 	}
 
-	udma_jfr->rq.ctx = udma_ctx;
 	if (udma_u_create_rq(udma_ctx, udma_jfr)) {
 		UDMA_LOG_ERR("failed to create jfr rqe buf.\n");
 		goto err_create_rq;
diff --git a/src/urma/hw/udma/udma_u_ops.c b/src/urma/hw/udma/udma_u_ops.c
index 8c89248..237284e 100644
--- a/src/urma/hw/udma/udma_u_ops.c
+++ b/src/urma/hw/udma/udma_u_ops.c
@@ -113,6 +113,7 @@ static void udma_u_init_context(struct udma_u_context *udma_ctx,
 	udma_ctx->dump_aux_info = resp->dump_aux_info;
 	udma_ctx->jfr_sge = resp->jfr_sge;
 	udma_ctx->hugepage_enable = resp->hugepage_enable;
+	udma_ctx->sva_sep_mode_en = resp->sva_sep_mode_en;
 }
 
 static void udma_u_destroy_jt_table(struct udma_u_context *udma_u_ctx)
diff --git a/src/urma/hw/udma/udma_u_segment.c b/src/urma/hw/udma/udma_u_segment.c
index c621d83..2dbdf72 100644
--- a/src/urma/hw/udma/udma_u_segment.c
+++ b/src/urma/hw/udma/udma_u_segment.c
@@ -95,7 +95,7 @@ static int udma_exec_register_seg_cmd(urma_context_t *urma_ctx,
 	urma_cmd_udrv_priv_t udata = {};
 	int ret;
 
-	if (seg_cfg->flag.bs.non_pin)
+	if (!to_udma_u_ctx(urma_ctx)->sva_sep_mode_en && seg_cfg->flag.bs.non_pin)
 		return 0;
 
 	ret = urma_cmd_register_seg(urma_ctx, urma_tseg, seg_cfg, &udata);
@@ -109,8 +109,8 @@ static int udma_u_check_seg_cfg(const urma_seg_cfg_t *seg_cfg)
 {
 	if (seg_cfg->flag.bs.access >= UDMA_SEGMENT_ACCESS_GUARD ||
 	    !seg_cfg->token_id || seg_cfg->flag.bs.token_policy >= URMA_TOKEN_SIGNED) {
-		UDMA_LOG_ERR("invalid segment input, access = %d, "
-			     "token_policy = %d, or NULL tid.\n",
+		UDMA_LOG_ERR("invalid segment input, access = %u, "
+			     "token_policy = %u, or NULL tid.\n",
 			     seg_cfg->flag.bs.access, seg_cfg->flag.bs.token_policy);
 		return EINVAL;
 	}
-- 
2.28.0.windows.1

